File: _java-archive/acquisition-service/src/main/java/dev/kritchalach/acquisition/service/AcquisitionService.java
package dev.kritchalach.acquisition.service;

import dev.kritchalach.acquisition.domain.Acquisition;
import dev.kritchalach.acquisition.domain.PropertySurvey;
import dev.kritchalach.acquisition.domain.SellerInfo;
import dev.kritchalach.acquisition.domain.WillingContract;
import dev.kritchalach.acquisition.event.AcquisitionApprovalRequestedEvent;
import dev.kritchalach.acquisition.event.AcquisitionContractDraftedEvent;
import dev.kritchalach.acquisition.event.AcquisitionApprovedEvent;
import dev.kritchalach.acquisition.event.PropertySurveyedEvent;
import dev.kritchalach.acquisition.repository.AcquisitionRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.UUID;

/**
 * Application Service สำหรับ Flow 1: Acquire Property
 * 1. รับ property.surveyed → สร้าง Acquisition + ขอ CEO approval
 * 2. รับ acquisition.approved → ร่าง WillingContract + emit acquisition.contract.drafted
 */
@Service
public class AcquisitionService {

    private static final Logger log = LoggerFactory.getLogger(AcquisitionService.class);

    private final AcquisitionRepository acquisitionRepository;

    public AcquisitionService(AcquisitionRepository acquisitionRepository) {
        this.acquisitionRepository = acquisitionRepository;
    }

    /**
     * Command: ReceiveSurvey + RequestApproval
     */
    public AcquisitionApprovalRequestedEvent receiveSurvey(PropertySurveyedEvent event) {
        UUID surveyId = UUID.fromString(event.getSurveyId());
        UUID propertyId = UUID.fromString(event.getPropertyId());

        log.info("Command: ReceiveSurvey — surveyId={}, propertyId={}", surveyId, propertyId);

        UUID sellerId = event.getSellerId() != null
                ? UUID.fromString(event.getSellerId())
                : UUID.randomUUID();

        PropertySurvey survey = new PropertySurvey(
                event.getAddress(), event.getAreaSqm(),
                event.getEstimatedValue(), event.getZoneType());

        SellerInfo seller = new SellerInfo(
                sellerId, event.getSellerName(), event.getSellerContact());

        Acquisition acquisition = Acquisition.fromSurvey(surveyId, propertyId, survey, seller);
        acquisition.requestApproval();
        acquisition = acquisitionRepository.save(acquisition);

        log.info("AcquisitionApprovalRequested: acquisitionId={}, status={}",
                acquisition.getAcquisitionId(), acquisition.getStatus());

        return new AcquisitionApprovalRequestedEvent(
                acquisition.getAcquisitionId(),
                acquisition.getSurveyId(),
                acquisition.getPropertyId(),
                survey.getAddress(),
                survey.getEstimatedValue(),
                seller.getSellerName(),
                acquisition.getCreatedAt()
        );
    }

    /**
     * Command: ApproveAcquisition + DraftWillingContract
     */
    public AcquisitionContractDraftedEvent draftContractAfterApproval(AcquisitionApprovedEvent event) {
        UUID acquisitionId = UUID.fromString(event.getAcquisitionId());

        log.info("Command: ApproveAcquisition + DraftWillingContract — acquisitionId={}", acquisitionId);

        Acquisition acquisition = acquisitionRepository.findById(acquisitionId)
                .orElseThrow(() -> new IllegalArgumentException(
                        "Acquisition not found: " + acquisitionId));

        acquisition.approve();

        UUID templateId = UUID.randomUUID();
        String fileUrl = "https://storage.realestate.com/willing-contracts/draft-"
                + UUID.randomUUID() + ".pdf";
        BigDecimal agreedPrice = event.getApprovedPrice() != null
                ? event.getApprovedPrice()
                : acquisition.getSurvey().getEstimatedValue();

        WillingContract willingContract = acquisition.draftWillingContract(
                templateId, fileUrl, agreedPrice);

        acquisition = acquisitionRepository.save(acquisition);

        log.info("AcquisitionContractDrafted: acquisitionId={}, willingContractId={}",
                acquisition.getAcquisitionId(), willingContract.getWillingContractId());

        return new AcquisitionContractDraftedEvent(
                acquisition.getAcquisitionId(),
                willingContract.getWillingContractId(),
                acquisition.getPropertyId(),
                willingContract.getFileUrl(),
                willingContract.getAgreedPrice(),
                willingContract.getTemplateId(),
                willingContract.getDraftedAt()
        );
    }
}

File: acquisition-service/src/service/AcquisitionService.ts
import { v4 as uuidv4 } from 'uuid';
import { Acquisition } from '../domain/Acquisition.js';
import { PropertySurvey } from '../domain/PropertySurvey.js';
import { SellerInfo } from '../domain/SellerInfo.js';
import { AcquisitionRepository } from '../repository/AcquisitionRepository.js';
import { coerceToUuid } from '../util/idCoerce.js';

export type { PropertySurveyedEvent } from '../event/PropertySurveyedEvent.js';
import type { PropertySurveyedEvent } from '../event/PropertySurveyedEvent.js';

export interface AcquisitionApprovalRequestedEvent {
  acquisitionId: string;
  surveyId: string;
  propertyId: string;
  address: string;
  estimatedValue: number;
  sellerName: string;
  requestedAt: string;
}

export interface AcquisitionApprovedEvent {
  acquisitionId: string;
  approvedPrice?: number;
  approvedBy?: string;
}

export interface AcquisitionContractDraftedEvent {
  acquisitionId: string;
  willingContractId: string;
  propertyId: string;
  fileUrl: string;
  agreedPrice: number;
  templateId: string;
  draftedAt: string;
}

export const AcquisitionService = {
  /**
   * Command: ReceiveSurvey + RequestApproval
   */
  async receiveSurvey(event: PropertySurveyedEvent): Promise<AcquisitionApprovalRequestedEvent> {
    console.log(
      `[Command] ReceiveSurvey — propertyId=${event.propertyId}, name=${event.propertyName ?? 'n/a'}, price=${event.price ?? 'n/a'}`,
    );

    // Coerce string codes (e.g. CEO sends "PROP-001") to deterministic UUIDs
    const surveyId = coerceToUuid(event.surveyId ?? event.propertyId);
    const propertyId = coerceToUuid(event.propertyId);
    const sellerId = event.sellerId ? coerceToUuid(event.sellerId) : uuidv4();

    // Backfill from new schema where fields differ in name
    const address = event.propertyAddress ?? event.address ?? '';
    const areaSqm = event.unitArea ?? event.areaSqm ?? 0;
    const estimatedValue = event.price ?? event.estimatedValue ?? 0;
    const zoneType = event.zoneType ?? event.propertyType ?? '';
    const sellerName = event.sellerName ?? event.propertyDeveloper ?? '';
    const sellerContact = event.sellerContact ?? '';

    const survey = new PropertySurvey(address, areaSqm, estimatedValue, zoneType);
    const seller = new SellerInfo(sellerId, sellerName, sellerContact);

    const acquisition = Acquisition.fromSurvey({
      surveyId,
      propertyId,
      survey,
      seller,
    });
    // Capture rich property + unit + pricing metadata
    acquisition.propertyDeveloper = event.propertyDeveloper;
    acquisition.propertyName = event.propertyName;
    acquisition.propertyType = event.propertyType;
    acquisition.propertyCode = event.propertyCode;
    acquisition.city = event.city;
    acquisition.currency = event.currency;
    acquisition.registration = event.registration;
    acquisition.createdBy = event.createdBy;
    acquisition.unitId = event.unitId ? coerceToUuid(event.unitId) : undefined;
    acquisition.unitCode = event.unitCode;
    acquisition.unitArea = event.unitArea;
    acquisition.bedroomType = event.bedroomType;
    acquisition.unitAddress = event.unitAddress;
    acquisition.bathrooms = event.bathrooms;
    acquisition.view = event.view;
    acquisition.furniture = event.furniture;
    acquisition.facility = event.facility;
    acquisition.pictureUrls = event.pictureUrls;
    acquisition.cost = event.cost;
    acquisition.minSalePrice = event.minSalePrice;
    acquisition.price = event.price;
    acquisition.saleTeamLead = event.saleTeamLead;
    acquisition.commission = event.commission;
    acquisition.externalStatus = event.status;

    acquisition.requestApproval();
    await AcquisitionRepository.save(acquisition);

    console.log(
      `[Domain Event] AcquisitionApprovalRequested: acquisitionId=${acquisition.acquisitionId}, status=${acquisition.status}`,
    );

    return {
      acquisitionId: acquisition.acquisitionId,
      surveyId: acquisition.surveyId,
      propertyId: acquisition.propertyId,
      address: survey.address,
      estimatedValue: survey.estimatedValue,
      sellerName: seller.sellerName,
      requestedAt: acquisition.createdAt,
    };
  },

  /**
   * Command: ApproveAcquisition + DraftWillingContract
   */
  async draftContractAfterApproval(event: AcquisitionApprovedEvent): Promise<AcquisitionContractDraftedEvent> {
    console.log(`[Command] DraftWillingContract — acquisitionId=${event.acquisitionId}`);

    const acquisition = await AcquisitionRepository.findById(event.acquisitionId);
    if (!acquisition) {
      throw new Error(`Acquisition not found: ${event.acquisitionId}`);
    }

    acquisition.approve();

    const templateId = uuidv4();
    const fileUrl = `https://storage.realestate.com/willing-contracts/draft-${uuidv4()}.pdf`;
    const agreedPrice = event.approvedPrice ?? acquisition.survey.estimatedValue;

    const willingContract = acquisition.draftWillingContract(templateId, fileUrl, agreedPrice);
    await AcquisitionRepository.save(acquisition);

    console.log(
      `[Domain Event] AcquisitionContractDrafted: acquisitionId=${acquisition.acquisitionId}, willingContractId=${willingContract.willingContractId}`,
    );

    return {
      acquisitionId: acquisition.acquisitionId,
      willingContractId: willingContract.willingContractId,
      propertyId: acquisition.propertyId,
      fileUrl: willingContract.fileUrl,
      agreedPrice: willingContract.agreedPrice,
      templateId: willingContract.templateId,
      draftedAt: willingContract.draftedAt,
    };
  },
};
